### PR TITLE
More date functionality

### PR DESCRIPTION
**Pull request by aditi-pandit**
i) CAST from string, timestamp
ii) PrestoSerializer code
iii) Greatest, Least, IN functions support
iv) BETWEEN dates support
v) DuckDB conversion
vi) ExpressionFuzzer/VectorFuzzer support
vii) Arrow conversion

### CODE REVIEW

**Comment by mbasmanova** [comment]
@aditi-pandit CI reports build failures. Would you take a look?

File: velox/expression/tests/CastExprTest.cpp [hunk 1]
```diff
@@ -320,6 +320,35 @@ TEST_F(CastExprTest, timestampAdjustToTimezoneInvalid) {
   EXPECT_THROW(testFunc(), std::runtime_error);
 }
 
+TEST_F(CastExprTest, date) {
+  testCast<std::string, Date>(
+      "date",
+      {
+          "1970-01-01",
```
Comment (mbasmanova) [inline velox/expression/tests/CastExprTest.cpp:327]: nit: perhaps, add some dates in the deep past and future? For example, "1812-04-15" and "2135-11-09".

File: velox/expression/tests/CastExprTest.cpp [hunk 2]
```diff
@@ -320,6 +320,35 @@ TEST_F(CastExprTest, timestampAdjustToTimezoneInvalid) {
...
+      {
+          Date(0),
+          Date(18262),
+          Date(-5),
+          Date(-18262),
+          std::nullopt,
+      });
+}
+
+TEST_F(CastExprTest, dateInvalid) {
```
Comment (mbasmanova) [inline velox/expression/tests/CastExprTest.cpp:342]: nit: invalidDate might read better

File: velox/expression/tests/CastExprTest.cpp [hunk 3]
```diff
@@ -320,6 +320,35 @@ TEST_F(CastExprTest, timestampAdjustToTimezoneInvalid) {
...
+  testCast<int8_t, Date>("date", {12}, {Date(0)}, true);
+  testCast<int16_t, Date>("date", {1234}, {Date(0)}, true);
+  testCast<int32_t, Date>("date", {1234}, {Date(0)}, true);
```
Comment (mbasmanova) [inline velox/expression/tests/CastExprTest.cpp:345]: perhaps, add an invalid date string, e.g. "2012-Oct-23"

File: velox/functions/prestosql/tests/ComparisonsTest.cpp [hunk 4]
```diff
@@ -52,3 +52,33 @@ TEST_F(ComparisonsTest, betweenVarchar) {
   EXPECT_EQ(true, between("orange"));
   EXPECT_EQ(true, between("pear"));
 }
+
+TEST_F(ComparisonsTest, betweenDate) {
+  Date beforeJune2019;
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ComparisonsTest.cpp:57]: This test would be shorter and easier to read if parseTo() took a string and returned a Date(). 

````
  std::vector<std::tuple<Date, bool>> testData = {
      {parse("2019-05-01"), false},
      {parse("2019-06-01"), true},
      ...};
```

File: velox/type/Conversions.h [hunk 5]
```diff
@@ -383,6 +383,38 @@ struct Converter<TypeKind::TIMESTAMP> {
   static T cast(const std::string& v, bool& nullOutput) {
     return fromTimestampString(v);
   }
+
+  static T cast(Date& d, bool& nullOutput) {
```
Comment (mbasmanova) [inline velox/type/Conversions.h:387]: d can be const &, right?

File: velox/type/Conversions.h [hunk 6]
```diff
@@ -383,6 +383,38 @@ struct Converter<TypeKind::TIMESTAMP> {
...
+    static const int32_t kMillisPerDay{86400000};
```
Comment (mbasmanova) [inline velox/type/Conversions.h:388]: nit: 86'400'000 is a bit easier to read

File: velox/type/Conversions.h [hunk 7]
```diff
@@ -383,6 +383,38 @@ struct Converter<TypeKind::TIMESTAMP> {
...
+  static T cast(const StringView& v, bool& nullOutput) {
+    return fromDateString(v);
+  }
+
+  static T cast(const std::string& v, bool& nullOutput) {
+    return fromDateString(v);
+  }
+
+  static T cast(Timestamp& t, bool& nullOutput) {
+    static const int32_t kSecsPerDay{86400};
```
Comment (mbasmanova) [inline velox/type/Conversions.h:415]: nit: kSecsPerDay -> kSecondsPerDay{86'400} for readability and to follow "don't abbreviate" guideline.

File: velox/type/tests/TypeTest.cpp [hunk 8]
```diff
@@ -162,6 +162,23 @@ TEST(Type, DateComparison) {
   EXPECT_GE(epoch, beforeEpoch);
 }
 
+TEST(Type, ParseStringToDate) {
```
Comment (mbasmanova) [inline velox/type/tests/TypeTest.cpp:165]: naming: ParseStringToDate -> parseStringToDate + let's file an issue to refactor this file to match coding guidelines

**Review by mbasmanova — COMMENTED** [review]
@aditi-pandit Looks great. Some minor comments below.

**Review by mbasmanova — APPROVED** [review]
Looks good to me. Thank you, Aditi.

File: velox/duckdb/conversion/tests/DuckWrapperTest.cpp [hunk 9]
```diff
@@ -159,9 +158,9 @@ TEST_F(BaseDuckWrapperTest, types) {
       {false, false, false, true});
 
   // date/timestamp
-  verifyUnaryResult<Timestamp>(
+  verifyUnaryResult<Date>(
       "SELECT i FROM (VALUES (DATE '1992-01-01'), (NULL)) tbl(i)",
-      {Timestamp(694224000, 0), Timestamp(0, 0)},
+      {Date(8035), Date(0)},
```
Comment (pedroerp) [inline velox/duckdb/conversion/tests/DuckWrapperTest.cpp:163]: nice :)

File: velox/type/Conversions.h [hunk 10]
```diff
@@ -383,6 +383,38 @@ struct Converter<TypeKind::TIMESTAMP> {
   static T cast(const std::string& v, bool& nullOutput) {
     return fromTimestampString(v);
   }
+
+  static T cast(const Date& d, bool& nullOutput) {
+    static const int32_t kMillisPerDay{86'400'000};
+    return Timestamp::fromMillis(d.days() * kMillisPerDay);
```
Comment (pedroerp) [inline velox/type/Conversions.h:389]: maybe we could have a `Timestamp::fromDays()` method?

File: velox/type/Conversions.h [hunk 11]
```diff
@@ -383,6 +383,38 @@ struct Converter<TypeKind::TIMESTAMP> {
...
+    return fromDateString(v);
+  }
+
+  static T cast(const std::string& v, bool& nullOutput) {
+    return fromDateString(v);
+  }
+
+  static T cast(const Timestamp& t, bool& nullOutput) {
+    static const int32_t kSecsPerDay{86'400};
+    return Date(t.getSeconds() / kSecsPerDay);
```
Comment (pedroerp) [inline velox/type/Conversions.h:416]: and a `Date::fromSeconds()`, for consistency? Even better, maybe a `Date::fromTimestamp()`

**Review by pedroerp — APPROVED** [review]
A few nits, but looks good!

File: velox/duckdb/conversion/tests/DuckWrapperTest.cpp (hunk 9 above)
Comment (aditi-pandit) [reply velox/duckdb/conversion/tests/DuckWrapperTest.cpp:163]: :)

File: velox/type/Conversions.h (hunk 10 above)
Comment (aditi-pandit) [reply velox/type/Conversions.h:389]: I will do a bit more cleanup in my next PR. Will address it there.

File: velox/type/Conversions.h (hunk 11 above)
Comment (aditi-pandit) [reply velox/type/Conversions.h:416]: I will do a bit more cleanup in my next PR. Will address it there.

**Comment by pedroerp** [comment]
@aditi-pandit our internal CI system detected a Fuzzer failure that sounds legit. Could you take a loot at it? You should be able to reproduce using something like:

> velox/expression/tests:velox_expression_fuzzer_test --seed 3031364857

This is a UBSAN failure, so we need to make sure you have it enabled to be able to reproduce it. This is the stack trace:

I1108 09:28:28.774264 19360 ExpressionFuzzer.cpp:494] ==============================> Started iteration 63 (seed: 3031364857)
I1108 09:28:28.774303 19360 ExpressionFuzzer.cpp:395] Couldn't find any function to return 'DATE'. Returning a constant instead.
I1108 09:28:28.774343 19360 ExpressionFuzzer.cpp:395] Couldn't find any function to return 'DATE'. Returning a constant instead.
velox/type/Date.cpp:28:30: runtime error: signed integer overflow: -1855961014 * 86400 cannot be represented in type 'int'
    #0 0x7f7a467af97b in facebook::velox::Date::toString[abi:cxx11]() const velox/type/Date.cpp:28
    #1 0x7f7a4693d91a in facebook::velox::variant::toJson[abi:cxx11]() const velox/type/Variant.cpp:288
    #2 0x7f7a4cdd659e in facebook::velox::core::ConstantTypedExpr::toString[abi:cxx11]() const velox/core/Expressions.h:91
    #3 0x7f7a4ce742dd in facebook::velox::core::CallTypedExpr::toString[abi:cxx11]() const velox/core/Expressions.h:182
    #4 0x7f7a4cdce18d in facebook::velox::test::(anonymous namespace)::ExpressionFuzzer::executeExpression(std::shared_ptr<facebook::velox::core::ITypedExpr const> const&, std::shared_ptr<facebook::velox::RowVector> const&, bool) velox/expression/tests/ExpressionFuzzer.cpp:420
    #5 0x7f7a4cd85811 in facebook::velox::test::(anonymous namespace)::ExpressionFuzzer::go(unsigned long) velox/expression/tests/ExpressionFuzzer.cpp:509
    #6 0x7f7a4cd84299 in facebook::velox::test::expressionFuzzer(std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<std::shared_ptr<facebook::velox::exec::FunctionSignature>, std::allocator<std::shared_ptr<facebook::velox::exec::FunctionSignature> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<std::shared_ptr<facebook::velox::exec::FunctionSignature>, std::allocator<std::shared_ptr<facebook::velox::exec::FunctionSignature> > > > > >, unsigned long, unsigned long) velox/expression/tests/ExpressionFuzzer.cpp:554
    #7 0x29bbe0 in main velox/expression/tests/ExpressionFuzzerTest.cpp:121
    #8 0x7f7a471b7d94 in __libc_start_main /home/engshare/third-party2/glibc/2.30/src/glibc-2.30/csu/../csu/libc-start.c:308:16
    #9 0x299029 in _start /home/engshare/third-party2/glibc/2.30/src/glibc-2.30/csu/../sysdeps/x86_64/start.S:120

SUMMARY: UndefinedBehaviorSanitizer: signed-integer-overflow velox/type/Date.cpp:28:30 in

**Comment by aditi-pandit** [comment]
> @aditi-pandit our internal CI system detected a Fuzzer failure that sounds legit. Could you take a loot at it? You should be able to reproduce using something like:
> 
> > velox/expression/tests:velox_expression_fuzzer_test --seed 3031364857
> 
> This is a UBSAN failure, so we need to make sure you have it enabled to be able to reproduce it. This is the stack trace:
> 
> I1108 09:28:28.774264 19360 ExpressionFuzzer.cpp:494] ==============================> Started iteration 63 (seed: 3031364857) I1108 09:28:28.774303 19360 ExpressionFuzzer.cpp:395] Couldn't find any function to return 'DATE'. Returning a constant instead. I1108 09:28:28.774343 19360 ExpressionFuzzer.cpp:395] Couldn't find any function to return 'DATE'. Returning a constant instead. velox/type/Date.cpp:28:30: runtime error: signed integer overflow: -1855961014 * 86400 cannot be represented in type 'int' #0 0x7f7a467af97b in facebook::velox::Date::toStringabi:cxx11 const velox/type/Date.cpp:28 #1 0x7f7a4693d91a in facebook::velox::variant::toJsonabi:cxx11 const velox/type/Variant.cpp:288 #2 0x7f7a4cdd659e in facebook::velox::core::ConstantTypedExpr::toStringabi:cxx11 const velox/core/Expressions.h:91 #3 0x7f7a4ce742dd in facebook::velox::core::CallTypedExpr::toStringabi:cxx11 const velox/core/Expressions.h:182 #4 0x7f7a4cdce18d in facebook::velox::test::(anonymous namespace)::ExpressionFuzzer::executeExpression(std::shared_ptr<facebook::velox::core::ITypedExpr const> const&, std::shared_ptrfacebook::velox::RowVector const&, bool) velox/expression/tests/ExpressionFuzzer.cpp:420 #5 0x7f7a4cd85811 in facebook::velox::test::(anonymous namespace)::ExpressionFuzzer::go(unsigned long) velox/expression/tests/ExpressionFuzzer.cpp:509 #6 0x7f7a4cd84299 in facebook::velox::test::expressionFuzzer(std::unordered_map<std::__cxx11::basic_string<char, std::char_traits, std::allocator >, std::vector<std::shared_ptrfacebook::velox::exec::FunctionSignature, std::allocator<std::shared_ptrfacebook::velox::exec::FunctionSignature > >, std::hash<std::__cxx11::basic_string<char, std::char_traits, std::allocator > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits, std::allocator > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits, std::allocator > const, std::vector<std::shared_ptrfacebook::velox::exec::FunctionSignature, std::allocator<std::shared_ptrfacebook::velox::exec::FunctionSignature > > > > >, unsigned long, unsigned long) velox/expression/tests/ExpressionFuzzer.cpp:554 #7 0x29bbe0 in main velox/expression/tests/ExpressionFuzzerTest.cpp:121 #8 0x7f7a471b7d94 in __libc_start_main /home/engshare/third-party2/glibc/2.30/src/glibc-2.30/csu/../csu/libc-start.c:308:16 #9 0x299029 in _start /home/engshare/third-party2/glibc/2.30/src/glibc-2.30/csu/../sysdeps/x86_64/start.S:120
> 
> SUMMARY: UndefinedBehaviorSanitizer: signed-integer-overflow velox/type/Date.cpp:28:30 in

Thanks Pedro ! Taking a look..

**Comment by pedroerp** [comment]
Looks like Fuzzer is still failing. @aditi-pandit , could take a look?

**Comment by aditi-pandit** [comment]
> Looks like Fuzzer is still failing. @aditi-pandit , could you take a look?

Thanks Pedro. Looking into it further.

**Comment by aditi-pandit** [comment]
> > Looks like Fuzzer is still failing. @aditi-pandit , could you take a look?
> 
> Thanks Pedro. Looking into it further.

Hi Pedro,
  The current code seems to have fixed the ExpressionFuzzer issue in the testing I did so far. Please can we check the merge again.

**Pull request closed without merging**